Sources/CacheAdvance/CacheAdvance.swift

```diff
@@ -0,0 +1,215 @@
+//
+//  Created by Dan Federman on 11/9/19.
+//  Copyright © 2019 Dan Federman.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+/// A cache that enables the performant persistence of individual messages to disk.
+/// This cache is intended to be written to and read from using the same serial queue.
+public final class CacheAdvance<T: Codable> {
+
+    // MARK: Initialization
+
+    /// Creates a new instance of the receiver.
+    ///
+    /// - Parameters:
+    ///   - file: The file URL indicating the desired location of the on-disk store. This file should already exist.
+    ///   - maximumBytes: The maximum size of the cache, in bytes. Logs larger than this size will fail to append to the store.
+    ///   - shouldOverwriteOldMessages: When `true`, once the on-disk store exceeds maximumBytes, new entries will replace the oldest entry.
+    ///
+    /// - Warning: `maximumBytes` must be consistent for the life of a cache. Changing this value after logs have been persisted to a cache leads to undefined behavior.
+    public init(
+        file: URL,
+        maximumBytes: Bytes,
+        shouldOverwriteOldMessages: Bool)
```

Comment:
I believe we spoke about this in person, but I can't remember what the answer was...
Let's say I create a `CacheAdvance` file with `shouldOverwriteOldMessages == true` and then I open it later with `shouldOverwriteOldMessages == false` (or vice versa).
How will we fail, if it all?
Is there a chance that we _think_ the data is valid and then read garbage data?

Reply:
The above scenario will lead to data loss. Byte 0 is always the beginning of `MessageSpan` followed by an encoded Message, so we'll never read garbage data. What will happen is that we'll fail to read the messages from the oldest message in the file until the end of the file, if the oldest message isn't at Byte 0.

I'll add a warning for this.